fix: oras pull error `empty response Docker-Content-Digest`

nima opened the pull request:
See https://github.com/oras-project/oras-go/issues/225 for details, but here's the **tldr**:

```
% oras pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Error: GET "https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag": empty response Docker-Content-Digest
```

The expected outcome (what this PR claims to fix):
```
% oras pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Downloading 3cd22ae067a3
Downloaded empty artifact
Pulled 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Digest: sha256:32aef68c20b6f25d69ecbc2f3f98e28f6f6fdacade75ee879459b61f610583f0
```

Resolves #225

nima commented:
### repro prior to the change:

```shell
➜  DoodleContainers git:(mainline) ✗ make
################################################################################
#------------------------------------------------------------------------------- oci:/home/nima/docker/config.json
aws --profile nima@ecr --region us-west-2 ecr get-login-password | docker login --username AWS --password-stdin 12345678901.dkr.ecr.us-west-2.amazonaws.com
WARNING! Your password will be stored unencrypted in /home/nima/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded

#------------------------------------------------------------------------------- oci:/home/nima/docker/config.json
aws --profile nima@ecr --region us-west-2 ecr get-login-password | oras login --username AWS --password-stdin 12345678901.dkr.ecr.us-west-2.amazonaws.com
Login Succeeded

#------------------------------------------------------------------------------- oci:setup
aws --profile nima@ecr --region us-west-2 ecr create-repository --repository-name my.repo --image-scanning-configuration scanOnPush=true | jq .
{
  "repository": {
    "repositoryUri": "12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo",
    "imageScanningConfiguration": {
      "scanOnPush": true
    },
    "encryptionConfiguration": {
      "encryptionType": "AES256"
    },
    "registryId": "12345678901",
    "imageTagMutability": "MUTABLE",
    "repositoryArn": "arn:aws:ecr:us-west-2:12345678901:repository/my.repo",
    "repositoryName": "my.repo",
    "createdAt": 1657590821
  }
}

#------------------------------------------------------------------------------- oci:setup
docker image build --tag my.repo:my.tag .
Sending build context to Docker daemon    276kB
Step 1/8 : FROM alpine:latest
 ---> e66264b98777
Step 2/8 : RUN mkdir -p /docked
 ---> Using cache
 ---> b70ff30822c5
Step 3/8 : WORKDIR /docked
 ---> Using cache
 ---> adbf13e74342
Step 4/8 : RUN apk add --no-cache socat
 ---> Using cache
 ---> 10673949eac4
Step 5/8 : EXPOSE 1234
 ---> Using cache
 ---> fe6e2db39608
Step 6/8 : EXPOSE 4321
 ---> Using cache
 ---> 69b73391ecf3
Step 7/8 : ENTRYPOINT [ "socat" ]
 ---> Using cache
 ---> 7ed365f9b91e
Step 8/8 : CMD [ "TCP4-LISTEN:1234", "TCP4-LISTEN:4321" ]
 ---> Using cache
 ---> cffeefdfa844
Successfully built cffeefdfa844
Successfully tagged my.repo:my.tag

#------------------------------------------------------------------------------- oci:setup
docker image tag my.repo:my.tag 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag

#------------------------------------------------------------------------------- oci:setup
docker image push 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
The push refers to repository [12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo]
336abb34c67c: Pushed
26a6b0d4e0cf: Pushed
24302eb7d908: Pushed
my.tag: digest: sha256:b9b2c380a97b8b4ab2acfd3d88850bc3ce2b95e803af33ac382e002b6683fbfd size: 945

#------------------------------------------------------------------------------- oci:main
docker image rm my.repo:my.tag
Untagged: my.repo:my.tag

#------------------------------------------------------------------------------- oci:main
oras pull 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Error: GET "https://12345678901.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag": empty response Docker-Content-Digest

make: *** [oci.main] Error 1
➜  DoodleContainers git:(mainline) ✗
```

### compiling the client against local lib...
```diff
git diff|cat
diff --git a/go.mod b/go.mod
index 958afdf..418ab8f 100644
--- a/go.mod
+++ b/go.mod
@@ -2,6 +2,8 @@ module oras.land/oras

 go 1.18

+replace oras.land/oras-go/v2 => /Users/nima/Development/github/oras-go
+
 require (
 	github.com/docker/cli v20.10.17+incompatible
 	github.com/moby/term v0.0.0-20210619224110-3f7ff695adc6
```

### run test again...
```shell
➜  DoodleContainers git:(mainline) ✗ make
################################################################################
#------------------------------------------------------------------------------- oci:/home/nima/docker/config.json
aws --profile nima@ecr --region us-west-2 ecr get-login-password | docker login --username AWS --password-stdin 12345678901.dkr.ecr.us-west-2.amazonaws.com
WARNING! Your password will be stored unencrypted in /home/nima/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded

#------------------------------------------------------------------------------- oci:/home/nima/docker/config.json
aws --profile nima@ecr --region us-west-2 ecr get-login-password | oras login --username AWS --password-stdin 12345678901.dkr.ecr.us-west-2.amazonaws.com
Login Succeeded

#------------------------------------------------------------------------------- oci:setup
aws --profile nima@ecr --region us-west-2 ecr create-repository --repository-name my.repo --image-scanning-configuration scanOnPush=true | jq .
{
  "repository": {
    "repositoryUri": "12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo",
    "imageScanningConfiguration": {
      "scanOnPush": true
    },
    "encryptionConfiguration": {
      "encryptionType": "AES256"
    },
    "registryId": "12345678901",
    "imageTagMutability": "MUTABLE",
    "repositoryArn": "arn:aws:ecr:us-west-2:12345678901:repository/my.repo",
    "repositoryName": "my.repo",
    "createdAt": 1657595026
  }
}

#------------------------------------------------------------------------------- oci:setup
docker image build --tag my.repo:my.tag .
Sending build context to Docker daemon    276kB
Step 1/8 : FROM alpine:latest
 ---> e66264b98777
Step 2/8 : RUN mkdir -p /docked
 ---> Using cache
 ---> b70ff30822c5
Step 3/8 : WORKDIR /docked
 ---> Using cache
 ---> adbf13e74342
Step 4/8 : RUN apk add --no-cache socat
 ---> Using cache
 ---> 10673949eac4
Step 5/8 : EXPOSE 1234
 ---> Using cache
 ---> fe6e2db39608
Step 6/8 : EXPOSE 4321
 ---> Using cache
 ---> 69b73391ecf3
Step 7/8 : ENTRYPOINT [ "socat" ]
 ---> Using cache
 ---> 7ed365f9b91e
Step 8/8 : CMD [ "TCP4-LISTEN:1234", "TCP4-LISTEN:4321" ]
 ---> Using cache
 ---> cffeefdfa844
Successfully built cffeefdfa844
Successfully tagged my.repo:my.tag

#------------------------------------------------------------------------------- oci:setup
docker image tag my.repo:my.tag 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag

#------------------------------------------------------------------------------- oci:setup
docker image push 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
The push refers to repository [12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo]
336abb34c67c: Pushed
26a6b0d4e0cf: Pushed
24302eb7d908: Pushed
my.tag: digest: sha256:b9b2c380a97b8b4ab2acfd3d88850bc3ce2b95e803af33ac382e002b6683fbfd size: 945

#------------------------------------------------------------------------------- oci:main
docker image rm my.repo:my.tag
Untagged: my.repo:my.tag

#------------------------------------------------------------------------------- oci:main
oras pull 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Downloaded empty artifact
Pulled 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Digest:

#------------------------------------------------------------------------------- oci:teardown
docker image rm my.repo:my.tag
Error: No such image: my.repo:my.tag

make: [oci.teardown] Error 1 (ignored)
#------------------------------------------------------------------------------- oci:teardown
docker image rm 12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Error response from daemon: conflict: unable to remove repository reference "12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag" (must force) - container 799004d623e6 is using its referenced image cffeefdfa844

make: [oci.teardown] Error 1 (ignored)
#------------------------------------------------------------------------------- oci:teardown
aws --profile nima@ecr --region us-west-2 ecr batch-delete-image --repository-name my.repo --image-ids imageTag=my.tag | jq .
{
  "failures": [],
  "imageIds": [
    {
      "imageTag": "my.tag",
      "imageDigest": "sha256:b9b2c380a97b8b4ab2acfd3d88850bc3ce2b95e803af33ac382e002b6683fbfd"
    }
  ]
}

#------------------------------------------------------------------------------- oci:teardown
aws --profile nima@ecr --region us-west-2 ecr delete-repository --repository-name my.repo | jq .
{
  "repository": {
    "repositoryUri": "12345678901.dkr.ecr.us-west-2.amazonaws.com/my.repo",
    "registryId": "12345678901",
    "imageTagMutability": "MUTABLE",
    "repositoryArn": "arn:aws:ecr:us-west-2:12345678901:repository/my.repo",
    "repositoryName": "my.repo",
    "createdAt": 1657595026
  }
}

GUIDE            : https://www.freecodecamp.org/news/the-docker-handbook/
DEMO             : oci
TARGET           : oci
ORAS_CLIENT_PATH : /home/nima/bin/oras
ORAS_CLIENT      : 0.13.0+unreleased
DOCKER_CLIENT    : 20.10.7
DOCKER_SERVER    : 20.10.7
################################################################################
➜  DoodleContainers git:(mainline) ✗
```

qweeah commented on internal/ioutil/io.go:
```diff
@@ -36,16 +36,19 @@ func (fn CloserFunc) Close() error {
+	if desc.Digest != "" {
+		verifier := desc.Digest.Verifier()
+		r = io.TeeReader(r, verifier)
+		if !verifier.Verified() {
```
I think `verifier.Verified()` should not be called before `io.ReadFull(r, buf);` is done. After `ReadFull` is done, the content stream is read (in the direction of below diagram) and digest can be properly calcaulated. cc @Wwwsylvia Can you help confirm?
```
            (r)                    (r)
+-----+ input reader +-----> output reader +-------->
                                +
                                |
                                +
                             verifier +----------->
```

nima replied:
Thanks for taking a look @qweeah; after going through this with @michaelb990 today, we realized there is actually another problem here.  At no point are we calculating the digest in `oras-go`, but instead, trusting what the server happens to return.  In the general case, that's probably fine, but I think the better approach is to use the digest returned by the server only for additional validation, as opposed to using it as the source-of-truth.

I believe what you're referring to is in reference to `TeeReader` which imposes a restriction of the read order, is that correct?

qweeah replied:
Yes. The `TeeReader` requires reader `r` to be fully read so the content can be written to the writer `verifier`.

Wwwsylvia commented:
Thanks @nima for the PR! Please let us know if this is ready for review. 😀

nima commented:
> Thanks @nima for the PR! Please let us know if this is ready for review. 😀

Thanks for taking a look @Wwwsylvia -- I have another PR out now, it's not intended to be approved, only to be inspected because I'm still not sure how to deal with the current issue.  I'm maintaining a log of my thoughts in [a thread](https://cloud-native.slack.com/archives/CJ1KHJM5Z/p1657935407555609) in the `#oras` slack channel, and also in the `git commit` message body.

nima commented:
`[INFO]` Current [slack thread](https://cloud-native.slack.com/archives/CJ1KHJM5Z/p1660767471849059) for this round.

nima commented:
I've set this to R4R despite the failure to expose the PR.  The failure is due to the fact that one of the dependencies of this package, `go-digest`, has also been modified as part of this change.  Locally, I've been using the`replace` directive in `go.mod` to point it to the locally modified version, but I'm not sure how to go about it here, other than get that change separately approved and merged in first, then continue with this.

shizhMSFT commented:
Few general comments to make this PR better.
- Please follow [conventional commits](https://www.conventionalcommits.org/) for PR title and body.
- Please [link this PR to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
- It is better to have a ruler at 80 chars for comment blocks so that people with small screens can read the code easily.
- Try not to modify the code of dependencies. If it is really needed, try proposing and merging the required changes to the upstream first.

nima commented:
Thanks for the pointers Shiwei, the first doc is specially useful.  I can retrospectively update my commit messages.

As for the second item, I've used the branch name `issues/225`; trying to do it the [correct way manually](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue), I don't seem to get the gear icon and unable to make the connection.

The third point, I believe I have done that already; can you point me to where I haven't?

And finally the last point, yes I'll keep that in mind; in this case I believe the change is still required; further clarity will no doubt come out at review time.

FeynmanZhou commented:
> Few general comments to make this PR better.
> 
> * Please follow [conventional commits](https://www.conventionalcommits.org/) for PR title and body.
> * Please [link this PR to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
> * It is better to have a ruler at 80 chars for comment blocks so that people with small screens can read the code easily.
> * Try not to modify the code of dependencies. If it is really needed, try proposing and merging the required changes to the upstream first.

The ORAS community also has [a simple version of Git Conventions](https://github.com/oras-project/community/blob/main/GIT_CONVENTIONS.md) for references.

michaelb990 commented on registry/remote/repository.go:
```diff
@@ -992,16 +1046,33 @@ func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
+// digest)
+func verifyContentDigests(actualDigest digest.Digest, digests map[string]digest.Digest) error {
+	for src, d := range digests {
+		if len(d.String()) == 0 {
```
why use an empty string and not `nil`?

nima replied:
`d` is a `digest.Digest`, which itself is a `string`.  This is why I have the other PR open in `go-digest`; a `d.IsEmpty()` seems better than `len(d.String()) == 0` or `d.String == ""`.

michaelb990 replied:
Okay. But why wouldn't we just let it be `nil`?

nima replied:
Because you can't set a string to a `nil`; it's not a `*digest.Digest` (pointer). Am I missing something?

michaelb990 commented on registry/reference.go:
```diff
@@ -56,33 +56,51 @@ type Reference struct {
 		return Reference{}, fmt.Errorf("%w: missing repository", errdef.ErrInvalidReference)
 	}
 	registry, path := parts[0], parts[1]
+	/*
+		The parameter `image` can thus be parsed (lowercase implies opaque, uppercase implies resolute)
+			<---------------- image ---------------------------------------------------> | Decode `image`
+			<=== REGISTRY ===> / <------------------ path -----------------------------> |  - Decode `path``
+			<=== REGISTRY ===> / <=== REPOSITORY ===> * <---------- reference ---------> |    - Decode `reference`
+			<=== REGISTRY ===> / <=== REPOSITORY ===> @ <=================== DIGEST ===> |      - Valid Form A
+			<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ===> @ <=== DIGEST ===> |      - Valid Form B
+			<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ======================> |      - Valid Form C
+			<=== REGISTRY ===> / <=== REPOSITORY ======================================> |    - Valid Form D
+
+		Note: DIGEST is itself, if present, of the form
+			<=========== DIGEST ==========>
+			<=== ALGO ===> : <=== HASH ===>
+
+		Note: Furthermore, in the case of Valid Form B, the TAG is dropped without any validation or consideration
+			  whatsoever.
+	*/
```
can we add unit tests for some of these cases instead of a comment?

nima replied:
I think the comment still helps, this polymorphic token deserves a docstring to show all the possible ways it can be used.  Not sure if the placement of this doc is the best though, maybe it can move somewhere else that's more appropriate.

michaelb990 replied:
Maybe I should rephrase this... can we add unit tests for this instead of *just* a comment?

nima replied:
Given that this function hasn't changed in any meaningful way, only annotated, why do you think it needs tests as part of this change?  I don't think it's a bad idea, just trying to see what made you see that as a blocker.

I'll take a look at adding tests meanwhile, maybe it will be quick.

nima replied:
Let me know if you think differently, but all 4 valid forms, as well as the invalid forms are already being tested.  I've annotated them in a change which I'll post in a bit.

nima replied:
[Done](https://github.com/oras-project/oras-go/pull/237/commits/150d0c325d5913eb96237d35bb5bc76fa5c6268f)

nima replied:
Will now look at where best to put that docstring.

nima replied:
[Done](https://github.com/oras-project/oras-go/pull/237/commits/833528e0ec96ee3c8909361cc85263bf40aaade2)

michaelb990 commented on registry/remote/repository.go:
```diff
@@ -895,7 +896,7 @@ func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.
 	switch resp.StatusCode {
 	case http.StatusOK:
-		return s.generateDescriptor(resp, ref)
+		return s.generateDescriptor(resp, ref, req.Method != "HEAD")
```
just for my knowledge, if the request method is `HEAD`, will the server always return the digest in the header? what if the request is for a tag and the server doesn't return the `Docker-Content-Digest` header? we wouldn't be able to generate a descriptor at all in that case, would we?

nima replied:
According to these test results, it will return that header in both cases; i.e., regardless of whether `image` is composed of a `tag`, or `digest`.  This is where I get thin on confidence though, so maybe you're seeing something I'm not.

The 4 cases below are labelled; 2 x pair of request/response, first via `digest` reference, second via `tag` reference.

```
% make main
[INFO] sourcing makefile.d/issue-225.mk
[INFO] sourced makefile.d/issue-225.mk
# >>> Request HTTP Headers [Digest Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840 -D- -o /dev/null
HTTP/1.1 200 OK
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT
Content-Length: 945


# <<< Response HTTP Headers [Digest Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840 -I
HTTP/1.1 200 OK
Content-Length: 945
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Content-Digest: sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT


# >>> Request HTTP Headers [Tag Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag -D- -o/dev/null
HTTP/1.1 200 OK
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT
Content-Length: 945


# <<< Response HTTP Headers [Tag Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag)
#------------------------------------------------------------------------------- issue-225:main
% curl -sL --user AWS:${AWS_PWD} https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag -I
HTTP/1.1 200 OK
Content-Length: 945
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Content-Digest: sha256:c84f0f11f3cdc4066362dc830e59f9d4b6a422eaaa94c2df1d3e7bfcb0384840
Docker-Distribution-Api-Version: registry/2.0
Sizes:
Date: Thu, 18 Aug 2022 19:25:43 GMT
```

nima replied:
I'm a bit confused here myself, I raised this question internally (20th of July) but haven't discovered anything new:
> Hmm, looks like the digest is returned from the server even if the reference is a tag (containing no digest)…

Now I'm trying to see exactly what HTTP call `oras pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag` translates to which results in the header to be missing and the error `empty response Docker-Content-Digest`.

@Wwwsylvia do you know of an easier way to translate an oras call to its component HTTP request?   Meanwhile I can see if I can run the `cli` through a proxy of some sort and inspect the HTTP request headers that way.

nima replied:
Okay here's a less ambiguous repro (part 1 of 2 from my last comment):
```
% make main
including makefile.d/headers.mk
included makefile.d/headers.mk
# HTTP Requests & Responses via [Tag Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag)
# >>> HEAD Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> HEAD /v2/my.repo/manifests/my.tag HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> HEAD Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Length: 1151
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Content-Digest: sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:25 GMT
<

# >>> GET Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> GET /v2/my.repo/manifests/my.tag HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> GET Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:26 GMT
< Content-Length: 1151
<

#
# HTTP Requests & Responses via [Digest Reference](https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e)
# >>> HEAD Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> HEAD /v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> HEAD Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sILo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Length: 1151
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Content-Digest: sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:26 GMT
<

# >>> GET Request
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[>]" | sed -E "s/(Authorization:.*) .*/\1 <REDACTED>/"
> GET /v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e HTTP/1.1
> Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
> Authorization: Basic <REDACTED>
> User-Agent: curl/7.79.1
> Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json
>

# >>> GET Response
#------------------------------------------------------------------------------- headers:main
% 2>/dev/null curl -sLo /dev/null -H "Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json" --trace-ascii curl.trace --user AWS:${AWS_PWD} -v https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e |& grep -E "^[<]"
< HTTP/1.1 200 OK
< Content-Type: application/vnd.docker.distribution.manifest.v2+json
< Docker-Distribution-Api-Version: registry/2.0
< Sizes:
< Date: Sat, 20 Aug 2022 00:14:27 GMT
< Content-Length: 1151
<

#
# Repro of issues/225 (HTTP GET)
#------------------------------------------------------------------------------- headers:main
% oras-main pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
########################## REQ #################################
GET /v2/my.repo/manifests/my.tag HTTP/1.1
Host: 080565210187.dkr.ecr.us-west-2.amazonaws.com
Accept: application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.cncf.oras.artifact.manifest.v1+json


########################## RES #################################
HTTP/1.1 200 OK
Content-Length: 1151
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Date: Sat, 20 Aug 2022 00:14:29 GMT
Docker-Distribution-Api-Version: registry/2.0
Sizes:


################################################################
Error: GET "https://080565210187.dkr.ecr.us-west-2.amazonaws.com/v2/my.repo/manifests/my.tag": empty response Docker-Content-Digest

make: *** [main] Error 1
%
```

Wwwsylvia replied:
Got it!

michaelb990 reviewed (commented):
thanks @nima! added some questions below.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -992,16 +1046,33 @@ func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
+// digest)
+func verifyContentDigests(actualDigest digest.Digest, digests map[string]digest.Digest) error {
+	for src, d := range digests {
+		if len(d.String()) == 0 {
```
As `digest.Digest` is a string, `if len(d) == 0` is more efficient.

nima replied:
This was more for communication to the reader, that while `d` may be a string, it's actually a `digest.Digest`.  If the change in `go-digest` I have open is shipped, we can instead use the abstract method `if d.IsEmpty`.  Not sure how much people on this project care about this abstraction that I'm talking about, since the counter argument is: "well we all know it's just a string".

I'll make it as per your suggestion meanwhile.  My preference would be `d.IsEmpty` because it is more communicative, but that's not available.  What are your thoughts?

Wwwsylvia replied:
Yeah it would be good if there is `d.IsEmpty`. But before there is one, I think we can just keep it simple.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -992,16 +1046,33 @@ func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
+		if len(d.String()) == 0 {
+			continue
+		}
+		if string(actualDigest) != string(d) {
```
`if actualDigest != d` would be enough.

nima replied:
Same as above, this was only for communication to the reader, to remind the reader that this `string` is actually a `digest.Digest`.  I'll change though, since preference is still for the `go-digest` change to go through, and then we can use `if d.IsIdenticalTo(actualDigest)`.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body, taking care not to destroy it in the
```
nit: Can we limit the length of one line of comment to 80 characters? It's the current convention of the oras-go repo.
We will need to have a contribution guide to call out all the conventions. /cc @shizhMSFT

nima replied:
Oh, I've been using line length of 120.  Should I make that change for all comments in the file then?

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 
 // generateDescriptor returns a descriptor generated from the response.
-func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference) (ocispec.Descriptor, error) {
+func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference, calculateDigest bool) (ocispec.Descriptor, error) {
```
When do we need to calculate the digest, and when do we not? Is it based on the request method?

nima replied:
Yes, based on the request method.  In the new revision (not yet published), I've updated `generateDescriptor` to take not the `bool calculateDigest`, but instead `string reqMethod`; then we do the "calculation" inside `generateDescriptor` and not have the caller make any decisions.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	switch {
+	case calculateDigest:
+		if respDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		var digests = map[string]digest.Digest{
+			"server-supplied": hdrDigest, // via response header `Docker-Content-Digest`
+			"client-supplied": refDigest, // via request reference string
+		}
+		if err = verifyContentDigests(respDigest, digests); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
+		}
+		dgst = respDigest
+	case refDigestErr == nil:
+		dgst = refDigest
+	case hdrDigestErr == nil:
+		dgst = hdrDigest
```
I'm just thinking out loud here:

1. If we need to calculate the digest
- -  Calculate the digest from the response, and compare it with the server-supplied value and client-supplied value

2. If we do not need to calculate the digest
- -  If the reference is a digest, then use it (client supplied) to generate the descriptor
- -  If the reference is not a digest, and the header contains a valid digest, then use the header digest (server supplied) to generate the descriptor
- - What if the reference is not a digest, and the header does not contain a valid digest? Will this case happen?

nima replied:
If that happens, in the `oras-go` lib, the codepath [here](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L994) will finish and the struct element `Digest` will remain as `""` in the return from that function.  I think what you might be asking though is, what is the implication of that.

It's a good question, I can mock it up in the `oras` client (drop `Docker-Content-Digest` from all server responses without discrimination of the call), and see how that behaves; that might leak some clues.

Wwwsylvia replied:
Got it.

Wwwsylvia replied:
> What if the reference is not a digest, and the digest header is not present? Would this case happen?

If this is not a valid case and should not happen, we should just return error.

Wwwsylvia replied:
Given that most registries return the `Docker-Content-Digest` header for `GET` requests, can the logic be simplified as following:

1. If `Docker-Content-Digest` header is present, calculating digest from content is not needed
- - If the client-supplied digest is available, verify the header digest against the client-supplied digest
- - If the client-supplied digest is not available, just respect the header digest

2. If `Docker-Content-Digest` is not present
- - If the request is a `GET`, then calculate the digest from the content. If the client-supplied digest is available, verify the content digest against the client-supplied digest; if the client-supplied digest is not available, just respect the content digest.
- - If the request is a `HEAD`: If the client-supplied digest is available, use client-supplied digest; if the client-supplied digest is not available, return error as we have no way to know the digest.

/cc @shizhMSFT What do you think?

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	switch {
+	case calculateDigest:
+		if respDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		var digests = map[string]digest.Digest{
+			"server-supplied": hdrDigest, // via response header `Docker-Content-Digest`
+			"client-supplied": refDigest, // via request reference string
+		}
+		if err = verifyContentDigests(respDigest, digests); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
+		}
+		dgst = respDigest
+	case refDigestErr == nil:
+		dgst = refDigest
+	case hdrDigestErr == nil:
+		dgst = hdrDigest
 	}
```
There is an edge case not caught by this logic and is a regression: What if a `HEAD` request to `/v2/<repo>/manifests/sha256:xxxxx` does not have a `Docker-Content-Digest` header in the response?

It is expected to have an error returned. However, the code logic will take the client supplied `sha256:xxxxx` as the source of the truth.

nima replied:
Thanks for the pointers Shiwei, the first doc is specially useful. I can retrospectively update my commit messages.

As for the second item, I've used the branch name issues/225; trying to do it the [correct way manually](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue), I don't seem to get the gear icon and unable to make the connection.

The third point, I believe I have done that already; can you point me to where I haven't?

And finally the last point, yes I'll keep that in mind; in this case I believe the change is still required; further clarity will no doubt come out at review time.

nima commented on registry/remote/repository.go:
```diff
@@ -895,7 +896,7 @@ func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.
 	switch resp.StatusCode {
 	case http.StatusOK:
-		return s.generateDescriptor(resp, ref)
+		return s.generateDescriptor(resp, ref, req.Method != "HEAD")
```
I think my understanding of this had some holes in it, it still does, but based on my new understanding let me refactor this.

nima commented on registry/remote/repository.go:
```diff
@@ -931,7 +932,7 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 	switch resp.StatusCode {
 	case http.StatusOK:
-		desc, err = s.generateDescriptor(resp, ref)
+		desc, err = s.generateDescriptor(resp, ref, req.Method != "HEAD")
```
Let me know if you think it's cleaner to simply forward `req.Method`, and the `generateDescriptor` decide; seems cleaner to me, and it's what I'll be publishing for the next commit in response of this PR, but I'm open to suggestions.

nima commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		return digest.Digest(""), err
+	}
+
+	var two bytes.Buffer
```
I've added in a comment which hopefully clarifies this, they're both basically interim buffers, `one` is used to calculate the digest, while the other (`two`) is used to replace the `resp.Body` which has been depleted as a result of the initial `read`.

Wwwsylvia replied:
Yeah but we can still have self-explainable names🤔. Naming is hard, but here is something I can think of:

- `one` -> `content`? `tr`?
- `two` -> `clonedBody`

nima commented on .gitignore:
```diff
@@ -10,6 +10,8 @@
 
 # Output of the go coverage tool, specifically when used with LiteIDE
 *.out
+.cover/
```
Ah right, I'll undo this change; thanks @shizhMSFT

nima commented on registry/reference.go:
```diff
@@ -56,33 +56,53 @@ type Reference struct {
-// If the reference contains both the tag and the digest, the tag will be
-// dropped.
+// ParseReference parses a string into an artifact reference.
+// If the reference contains both the tag and the digest, the tag will be dropped.
```
Yes, and done! Currently I've set code-wrap to be at 120; is that reasonable/adopted standard in this project?

nima commented on registry/reference.go:
```diff
@@ -56,33 +56,53 @@ type Reference struct {
+//
+//	Note: Furthermore, in the case of Valid Form B, the TAG is dropped without any validation or consideration
+//		  whatsoever.
+func ParseReference(image string) (Reference, error) {
```
An image is an artifact, but an artifact is not necessarily an image; this makes good sense.

nima commented on registry/reference.go:
```diff
@@ -56,33 +56,53 @@ type Reference struct {
-func ParseReference(raw string) (Reference, error) {
-	parts := strings.SplitN(raw, "/", 2)
+//
+// The parameter `image` is an opaque polymorphic string, so it's worth layout out the 4 valid forms that it can take.
+// This parameter can be parsed as follows; *lowercase* labels here imply opaque values (composed of other components),
+// and *uppercase* labels imply resolute values (can not be expanded any further).
+//
+//		<---------------- image ---------------------------------------------------> | Decode `image`
+//		<=== REGISTRY ===> / <------------------ path -----------------------------> |  - Decode `path``
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> * <---------- reference ---------> |    - Decode `reference`
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> @ <=================== DIGEST ===> |      - Valid Form A
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ===> @ <=== DIGEST ===> |      - Valid Form B
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ======================> |      - Valid Form C
+//		<=== REGISTRY ===> / <=== REPOSITORY ======================================> |    - Valid Form D
+//
+//	Note: DIGEST is itself, if present, of the form
+//		<=========== DIGEST ==========>
+//		<=== ALGO ===> : <=== HASH ===>
+//
+//	Note: Furthermore, in the case of Valid Form B, the TAG is dropped without any validation or consideration
+//		  whatsoever.
```
Good idea, BNF is much more compact so I can defined the entire range of possibilities.  I might leave the section of the existing docstring though, since it helps to map the forms to the 4x valid forms.  Let me know what you think once I publish it.

nima replied:
Oh, and the godoc looks reasonable to me, let me know if you would like a screenshot.

nima commented on registry/reference.go:
```diff
@@ -56,33 +56,53 @@ type Reference struct {
+//
+//	Note: Furthermore, in the case of Valid Form B, the TAG is dropped without any validation or consideration
+//		  whatsoever.
+func ParseReference(image string) (Reference, error) {
```
Sure I can do this, it won't need a new associated "issue" will it?

nima commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+
+	resp.Body = io.NopCloser(bytes.NewReader(two.Bytes()))
+
+	return d, err
```
Good catch; fixed.

nima commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		return d, fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+
+	resp.Body = io.NopCloser(bytes.NewReader(two.Bytes()))
```
Oh very nice, I think you mean `&two` though; fixed.

nima commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		if respDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		var digests = map[string]digest.Digest{
+			"server-supplied": hdrDigest, // via response header `Docker-Content-Digest`
+			"client-supplied": refDigest, // via request reference string
+		}
+		if err = verifyContentDigests(respDigest, digests); err != nil {
```
Don't think it helps readability? or worse, hinders it?  I can make two separate variables instead, let me know what you think and I'll make the change.

nima commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+			return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
+		}
+		dgst = respDigest
+	case refDigestErr == nil:
+		dgst = refDigest
+	case hdrDigestErr == nil:
+		dgst = hdrDigest
```
Fixed.

nima commented:
I've published a new PR; I have hopefully addressed all comments (which by the way were very helpful).

The only item I have not addressed in this PR, is to split this PR into two (one for the bug-fix, and one for changes to `reference*.go`).

I also have a few items that I have addressed, but are requesting further feedback; nothing major.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -944,37 +946,69 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
-	}
+	// HTTP HEAD Requests are equipped with the `Docker-Content-Digest` header.
+	// HTTP GET Requests do not.
```
HTTP GET Requests **may** not?
Some registries do return the `Docker-Content-Digest` header for Get requests.

Wwwsylvia replied:
Also, can we add unit tests for the cases where the `Docker-Content-Digest` header is not returned?

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+
+	// 3. Validate Digest
+	var dgst, respDigest, refDigest, hdrDigest digest.Digest
+	var refDigestErr, hdrDigestErr error
```
Agreed.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | FAIL                |
+	| 5  | validDigest   | presentValid   | TRUST,COMPARE,PASS    | TRUST,COMPARE,PASS  |
+	| 6  | validDigest   | presentInvalid | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 7  | invalidDigest | missing        | CALCULATE,VERIFY,FAIL | FAIL                |
+	| 8  | invalidDigest | presentValid   | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 9  | invalidDigest | presentInvalid | TRUST,COMPARE,!PASS   | TRUST,COMPARE,!PASS |
```
"Invalid client digest" is actually treated as tag. Therefore, we don't have case 7-9, do we?

nima replied:
Let me see if I understand this on a second look...  Let's look at the case where the server does not have the special header--given that the `ref` is syntactically correct, but otherwise invalid:

1. an "invalid" (syntactically correct digest, but an invalid one) comes in to `generateManifestDescriptor#L1010`
2. `->` [`oras-go:registry/remote/repository.go#L1010`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1010)
3. `->` [`oras-go:registry/reference.go#L160`](https://github.com/nima/oras-go/blob/issues/225/registry/reference.go#L160)
4. `->` [`go-digest:digest.go#L81`](https://github.com/opencontainers/go-digest/blob/master/digest.go#L81)
5. `->` [`go-digest:digest.go#L103`](https://github.com/opencontainers/go-digest/blob/master/digest.go#L103)
6. `->` [`go-digest:algorithm.go#L179`](https://github.com/opencontainers/go-digest/blob/master/algorithm.go#L179) (returns `nil`, since it's a "valid" digest)
7. `<-` Item `5` returns `nil` (no error) from `Validate`
8. `<-` Item `4` returns `<digest.Digest>, nil` from `Parse`
9. `<-` Item `3` returns `<digest.Digest>, nil` from `Digest`
10. [`oras-go:registry/remote/repository.go#L1011`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1011) sets `refDigest`
11. [`oras-go:registry/remote/repository.go#L1056`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1056) returns an error, due to the mismatch, for both `HEAD`, and `GET`

Now, let's look at a tag:

1. *same as above*
2. *same as above*
3. *same as above*
4. *same as above*
5. *same as above*
6. `<-` [`go-digest:digest.go#L107`](https://github.com/opencontainers/go-digest/blob/master/digest.go#L107) (`5`) returns `ErrDigestInvalidFormat`
7. `<-` `4` returns `<digest.Digest>, ErrDigestInvalidFormat` from `Parse`
8. `<-` Item `3` returns `<digest.DigestErrDigestInvalidFormat nil` from `Digest`
9. [`oras-go:registry/remote/repository.go#L1011`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1011) does not set `refDigest`
10. `HEAD` calls return an error @ [`oras-go:remote/repository.go#L1037`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1037-L1040)
11. `GET` callers get through unscathed @ [`oras-go:remote/repository.go#L1060`](https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1060-L1064)

Those seem different, so I'm trying to understand why you're suggesting that a `tag reference` is equivalent to a "invalid" `digest reference`.

nima replied:
Discussion in Slack has resolved this conversation.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	// 5a. Set contentDigest if no early error conditions are encountered
+	var calculatedDigest digest.Digest
+	if len(hdrDigest) == 0 {
+		if httpMethod == "HEAD" {
```
```suggestion
		if httpMethod == http.MethodHead {
```

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -41,6 +42,8 @@ import (
 	"oras.land/oras-go/v2/registry/remote/internal/errutil"
 )
 
+const dockerContentDigestHeader = "Docker-Content-Digest"
```
It is better to document this constant.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	// 5b. With all 3 digests appropriately set, perform final 3-way validation
+	if err = verifyContentDigests(calculatedDigest, hdrDigest, refDigest); err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
 	}
```
Do we need 3-way validation?

`calculatedDigest` has value if and only if `hdrDigest` is empty. Therefore, based on line 1047 and 1049, we only need to verify `contentDigest` and `refDigest`, which can be done easily as follows.

```go
if refDigest != "" && refDigest != contentDigest {
        return ocispec.Descriptor{}, fmt.Errorf("%s: mismatch digest: %s", refDigest, contentDigest)
}
```

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+
+	var d digest.Digest
+	var err error
+	body := limitReader(resp.Body, 0)
```
`MaxMetadataBytes` need to be honored for security considerations.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+// calculateDigestFromResponse calculates the actual digest of the response body
+// taking care not to destroy it in the process.
+func calculateDigestFromResponse(resp *http.Response) (digest.Digest, error) {
+	// Read from `resp.Body` and (via io.TeeReader), duplicate the read data
+	// into the new buffers.  This read will "deplete" resp.Body which, having
+	// no "rewind" button, will become useless.  At which point, one clone is
+	//used to calculate the digest, and another is used replace the now-defunct
+	// `resp.Body`.
+	var cloneForDigestCalculation io.Reader
+	var cloneForRestoration bytes.Buffer
+
+	var d digest.Digest
+	var err error
+	body := limitReader(resp.Body, 0)
+	cloneForDigestCalculation = io.TeeReader(body, &cloneForRestoration)
+	if d, err = digest.FromReader(cloneForDigestCalculation); err != nil {
+		return d, fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+
+	resp.Body = io.NopCloser(&cloneForRestoration)
+
+	return d, nil
```
Since we will read all of the response body into the memory anyway, why don't we do something like this?
```go
defer resp.Body.Close() // Yes, we need to close the body to free up resources like connections

body := limitReader(resp.Body, MaxMetadataBytes) // You need to find a way to pass this variable from `Repository` via `manifestStore` or `blobStore`
content, err := io.ReadAll(body)
if err != nil {
    return "", fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
}
resp.Body = io.NopCloser(bytes.NewReader(content))
return digest.FromBytes(content), nil
```

It is much simpler, cleaner, and faster.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+			"%s %q: invalid response %s: %s",
+			resp.Request.Method, resp.Request.URL, dockerContentDigestHeader, digestStr,
+		)
+	} else if contentDigest != expected {
```
```suggestion
	}
	if contentDigest != expected {
```
See https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

shizhMSFT commented on registry/remote/repository_test.go:
```diff
@@ -2210,6 +2210,98 @@ func Test_BlobStore_Resolve(t *testing.T) {
+	reference := registry.Reference{
+		Registry:   "eastern.haan.com",
+		Reference:  "<calculate>",
+		Repository: "25To220CE",
```
The repository name is invalid since all the chars should be in lower cases.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	bling, and not performing the expensive validation here for performance
+	reasons, knowing that it will be calculated at a later point in time.
+
+	 ___________________________________________________________________________________
+	| ID | CLIENT        | SERVER         | GET                   | HEAD                |
+	|----+---------------+----------------+-----------------------+---------------------+
+	| 1  | tag           | missing        | CALCULATE,PASS        | FAIL                |
+	| 2  | tag           | presentValid   | TRUST,PASS            | TRUST,PASS          |
+	| 3  | tag           | presentInvalid | TRUST,!PASS           | TRUST,!PASS         |
+	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | FAIL                |
+	| 5  | validDigest   | presentValid   | TRUST,COMPARE,PASS    | TRUST,COMPARE,PASS  |
+	| 6  | validDigest   | presentInvalid | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 7  | invalidDigest | missing        | CALCULATE,VERIFY,FAIL | FAIL                |
+	| 8  | invalidDigest | presentValid   | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 9  | invalidDigest | presentInvalid | TRUST,COMPARE,!PASS   | TRUST,COMPARE,!PASS |
+	 -----------------------------------------------------------------------------------
```
nit: nice table 🎉

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+//	| 8  | invalidDigest | presentValid   | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+//	| 9  | invalidDigest | presentInvalid | TRUST,COMPARE,!PASS   | TRUST,COMPARE,!PASS |
+//	 -----------------------------------------------------------------------------------
+func generateManifestDescriptor(resp *http.Response, ref registry.Reference, httpMethod string, maxMetadataBytes int64) (ocispec.Descriptor, error) {
```
Do we need to pass the `maxMetadataBytes` as a parameter?

nima replied:
How else would we be able to access it then?

Wwwsylvia replied:
Sorry I misunderstood something. But the suggestion in [this comment](https://github.com/oras-project/oras-go/pull/237#discussion_r953362639) looks nice.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+
+	if contentDigest, err := digest.Parse(digestStr); err == nil {
+		if contentDigest != expected {
+			return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
```
nit: Maybe we can elaborate this error message a bit?

nima replied:
Done.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+func calculateDigestFromResponse(resp *http.Response, maxMetadataBytes int64) (digest.Digest, error) {
+	defer resp.Body.Close()
+
+	body := limitReader(resp.Body, maxMetadataBytes)
```
I think we can use the `MaxMetadataBytes` here, instead of passing the value around. As it looks like we won't set other values.

nima replied:
I don't follow @Wwwsylvia, the only `MaxMetadataBytes` I can see is the declaration of the type on line 102.  Do you mean we can use the default (via `0`)?

shizhMSFT replied:
I think @Wwwsylvia is referring to make `calculateDigestFromResponse()` a method of `*manifestStore`.

However, the method `calculateDigestFromResponse()` looks like a common method, which can be re-used by others. Let's keep it as it is without making it a method of `*manifestStore`.

Wwwsylvia replied:
Sorry I misunderstood something, please ignore this one.

nima replied:
Lol; yeah it now makes sense to convert `generateManifestDescriptor` back to a method (`generateDescriptor`).  Done.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+//	| 1  | tag           | missing        | CALCULATE,PASS        | FAIL                |
+//	| 2  | tag           | presentValid   | TRUST,PASS            | TRUST,PASS          |
+//	| 3  | tag           | presentInvalid | TRUST,!PASS           | TRUST,!PASS         |
+//	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | FAIL                |
```
For row 4, it looks like the behavior is different for blobs and manifests. We need to keep them consistent.
When the client does `HEAD` request by digest, and the server header is missing,
1. `generateBlobDescriptor` honors the client digest
2. `generateManifestDescriptor` returns an error

In the original implementation, we honor the client digest for both blobs and manifest. See https://github.com/oras-project/oras-go/blob/main/registry/remote/repository.go#L958-L968.

nima replied:
That would mean:
```
//	 ___________________________________________________________________________________
//	| ID | CLIENT        | SERVER         | GET                   | HEAD                |
//	|----+---------------+----------------+-----------------------+---------------------+
//	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | PASS                |
 //	| 7  | invalidDigest | missing        | CALCULATE,VERIFY,FAIL | FAIL                |
```

What would happen to row 7 then? Or like @shizhMSFT suggested earlier, I think I should drop 7-9 altogether.

nima replied:
Also if they're to be in sync, then the tests should not have two variables `errExpectedOnHEAD` and `errExpectedOnGET`, rather just one `errExpected`.  Or the consistency you're referring to the case of row 4 only?

Wwwsylvia replied:
The consistency I referred to is between blob descriptors and manifest descriptors. It makes sense that `HEAD` and `GET` behaves differently, like row 1.

nima replied:
Very good point, and this is what prompted me to expand the truth table to include two new rows for Blobs.

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	}
+
+	// 3. Validate Client Reference
+	var refDigest digest.Digest
+	if d, err := ref.Digest(); err == nil {
+		refDigest = d
+	}
```
It looks like the `refDigest` is not used before line 1053. I think we can move this part there.

nima replied:
With the change concerning row 4, it's now used on 1035 (not yet published).

Wwwsylvia commented on registry/remote/repository_test.go:
```diff
@@ -2210,6 +2210,98 @@ func Test_BlobStore_Resolve(t *testing.T) {
 	}
 }
 
+func Test_Manifest_generateManifestDescriptorWithVariousDockerContentDigestHeaders(t *testing.T) {
```
If this is to test the `generateManifestDescriptor` method, we should call it ` Test_generateManifestDescriptor`.

Wwwsylvia replied:
Do we want to also test the `generateBlobDescriptor` method?

nima replied:
Yeah the name was bad; since I've made it a method `(s *manifestStore)` again (to access `s.repo.MaxMetadataBytes`, the test is now called `Test_ManifestStore_generateDescriptorWithVariousDockerContentDigestHeaders`

nima replied:
Let me take a look at `generateBlobDescriptor` and add some tests.

nima replied:
This is done and **`R4R`**

nima reviewed (commented):
I'll take another look at this in the morning, but have left some comments in the meanwhile.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -895,7 +903,7 @@ func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.
 	switch resp.StatusCode {
 	case http.StatusOK:
-		return s.generateDescriptor(resp, ref)
+		return generateManifestDescriptor(resp, ref, req.Method, s.repo.MaxMetadataBytes)
```
Looks like we can simplify this back to
```suggestion
		return s.generateDescriptor(resp, ref, req.Method)
```

nima replied:
Yes; Done.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 		return nil
 	}
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
-	}
-	if contentDigest != expected {
-		return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
+
+	if contentDigest, err := digest.Parse(digestStr); err == nil {
+		if contentDigest != expected {
+			return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
+		}
+	} else {
+		return fmt.Errorf(
+			"%s %q: invalid response %s: %s",
+			resp.Request.Method, resp.Request.URL, dockerContentDigestHeader, digestStr,
+		)
 	}
+
 	return nil
```
Comparing with the previous implementation, the logic is exactly the same but the readability is reduced.

Could we revert the change but retain the change of `dockerContentDigestHeader`?

nima replied:
Yes you're right; Done.

shizhMSFT commented on registry/remote/repository_test.go:
```diff
@@ -2210,6 +2210,98 @@ func Test_BlobStore_Resolve(t *testing.T) {
+	const theAmazingBanClan = "Ban Gu, Ban Chao, Ban Zhao"
+	const theAmazingBanDigest = "b526a4f2be963a2f9b0990c001255669eab8a254ab1a6e3f84f1820212ac7078"
+	correctDigest := fmt.Sprintf("sha256:%v", theAmazingBanDigest)
+	incorrectDigest := fmt.Sprintf("sha256:%v", "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
```
There is no concept of `incorrectDigest`. As long as the digest is in a right form, it is valid although it may not be desired. In that case, it is a customer error.

nima replied:
I'm having problems convincing myself of that; could you help me understand how?  I've got the two code-paths traversed [here](https://github.com/oras-project/oras-go/pull/237#discussion_r954289594).

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -776,8 +784,8 @@ func (s *blobStore) FetchReference(ctx context.Context, reference string) (desc
-// generateDescriptor returns a descriptor generated from the response.
-func (s *blobStore) generateDescriptor(resp *http.Response, refDigest digest.Digest) (ocispec.Descriptor, error) {
+// generateBlobDescriptor returns a descriptor generated from the response.
+func generateBlobDescriptor(resp *http.Response, refDigest digest.Digest) (ocispec.Descriptor, error) {
```
Yeah it's not about the change, it is just a general question. 😄

nima replied:
Got it... though I still don't understand the question I'm afraid.  `generateBlobDescriptor` doesn't *ever* return a `Docker-Content-Digest`, since it's simply returning a `ocispec.Descriptor`.  I'm sure that's not exactly what you meant, but I'm unable to determine what you *did* intend to ask.  The only header I see that function look at is `Content-Type`.

Wwwsylvia replied:
Sorry for being unclear. I actually mean, will the ECR server always return the `Docker-Content-Digest` header for blob `HEAD` and `GET` requests?

nima replied:
Ah, now I get you -- yes ECR will always return that header.

cc: @michaelb990

nima commented on registry/remote/repository.go:
```diff
@@ -931,7 +932,7 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 	switch resp.StatusCode {
 	case http.StatusOK:
-		desc, err = s.generateDescriptor(resp, ref)
+		desc, err = s.generateDescriptor(resp, ref, req.Method != "HEAD")
```
In the current revision, we calculate the digest only when necessary (`GET` and header is missing); no longer is it done indiscriminately.  The reason for this is performance, for anyone reading this without context.  I think that resolves this conversation.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+
+	// 4. Validate Server Digest (if present)
+	var serverHeaderDigest digest.Digest
+	if serverHeaderDigestErr := resp.Header.Get(dockerContentDigestHeader); serverHeaderDigestErr != "" {
+		if serverHeaderDigest, err = digest.Parse(serverHeaderDigestErr); err != nil {
```
Looks like it is a typo
```suggestion
	if serverHeaderDigestStr := resp.Header.Get(dockerContentDigestHeader); serverHeaderDigestErr != "" {
		if serverHeaderDigest, err = digest.Parse(serverHeaderDigestStr); err != nil {
```

nima replied:
Ouch, thanks; fixed!

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	var calculatedDigest digest.Digest
+	if len(serverHeaderDigest) == 0 {
+		if httpMethod == http.MethodHead {
+			if len(refDigest) == 0 {
+				// HEAD without server `Docker-Content-Digest` header is an
+				// immediate fail; HEAD
+				return ocispec.Descriptor{}, fmt.Errorf(
+					"HTTP %s request missing required header `%s`",
+					httpMethod, dockerContentDigestHeader,
+				)
+			}
```
This check raises an interesting case: What if the server does not return a content digest header but client has one?
It seems to be ok to trust it as it will be checked later at the client side.

Suggestions: updated doc and added `contentDigest = refDigest`.
```suggestion
			if len(refDigest) == 0 {
				// HEAD without server `Docker-Content-Digest` header is an
				// immediate fail
				return ocispec.Descriptor{}, fmt.Errorf(
					"HTTP %s request missing required header `%s`",
					httpMethod, dockerContentDigestHeader,
				)
			}
			contentDigest = refDigest
```

nima replied:
Done.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	if len(refDigest) > 0 && len(contentDigest) > 0 && refDigest != contentDigest {
```
We don't need to and should not check `len(contentDigest) > 0`. If `len(refDigest) > 0` is true but `len(contentDigest) == 0`, then we will always have `refDigest != contentDigest`.

Again, if `len(contentDigest) == 0`, the entire Boolean expression is false. Therefore, it passes the check, and we will return a descriptor with an empty digest, which is problematic.

nima replied:
Now since we're trusting the client, and setting `contentDigest = refDigest` as per https://github.com/oras-project/oras-go/pull/237#discussion_r954467389, yes I can again drop this change (this was just added in the last push).

Done.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	if contentDigest != expected {
-		return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
+
+	if len(expected) > 0 && contentDigest != expected {
```
I'm curious that in which situation we will pass an empty `expected`?

nima replied:
Removing that check, the following tests fail...
```
--- FAIL: Test_generateBlobDescriptorWithVariousDockerContentDigestHeaders (0.00s)
    repository_test.go:2576: [Blob.GET] 3. Client:Tag & Server:DigestWrongButSyntacticallyValid; expected no error for request, but got err: GET <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` vs expected ``
    repository_test.go:2576: [Blob.HEAD] 3. Client:Tag & Server:DigestWrongButSyntacticallyValid; expected no error for request, but got err: HEAD <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` vs expected ``
    repository_test.go:2576: [Blob.GET] 2. Client:Tag & Server:DigestValid; expected no error for request, but got err: GET <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:b526a4f2be963a2f9b0990c001255669eab8a254ab1a6e3f84f1820212ac7078` vs expected ``
    repository_test.go:2576: [Blob.HEAD] 2. Client:Tag & Server:DigestValid; expected no error for request, but got err: HEAD <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:b526a4f2be963a2f9b0990c001255669eab8a254ab1a6e3f84f1820212ac7078` vs expected ``
```

Going to have some food, then think about this a bit more.

shizhMSFT replied:
I think there's something wrong with the tests. Let me take a deeper look.

shizhMSFT replied:
I've checked the related code. The `refDigest` in `generateBlobDescriptor()` should never be empty. This is guaranteed by other checks before invoking `generateBlobDescriptor()`.
```go
func generateBlobDescriptor(resp *http.Response, refDigest digest.Digest)
```
Therefore, we should revise the related tests and remove the check of `len(expected) > 0`.

nima replied:
Okay, the code has been reverted (no check for `len(expected)` just for the sake of testing), and instead, the test has been updated to account for this otherwise impossible scenario.

nima reviewed (commented):
Okay I think I've addressed all comments so far.  Also expanded on what happens when `expected` is not length-checked.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,132 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	if len(serverHeaderDigest) == 0 {
+		if httpMethod == http.MethodHead {
+			if len(refDigest) == 0 {
+				// HEAD without server `Docker-Content-Digest` header is an
+				// immediate fail; HEAD
```
nit:
```suggestion
				// HEAD without server `Docker-Content-Digest` header is an
				// immediate fail
```

nima replied:
Oops; fixed!

Wwwsylvia commented on registry/remote/repository_test.go:
```diff
@@ -43,6 +43,106 @@ import (
 )
 
+// For the Truth Table that these tests are modeled on, see the following:
+// https://github.com/nima/oras-go/blob/issues/225/registry/remote/repository.go#L1009-L1030
```
nit: Is the truth table moved?

nima replied:
Yes!  Done.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+	var calculatedDigest digest.Digest
```
nice catch.

nima replied:
Yes! Done.

shizhMSFT commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,132 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		} else {
+			// GET without server `Docker-Content-Digest` header forces the
+			// expensive calculation
+			if calculatedDigest, err = calculateDigestFromResponse(resp, s.repo.MaxMetadataBytes); err != nil {
+				return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+			}
```
This suggestion reflects @Wwwsylvia 's [comment](https://github.com/oras-project/oras-go/pull/237/files#r954473588).

```suggestion
			calculatedDigest, err := calculateDigestFromResponse(resp, s.repo.MaxMetadataBytes)
			if err != nil {
				return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
			}
```

shizhMSFT replied:
with line 1001: `var calculatedDigest digest.Digest` removed

nima replied:
Fixed.

shizhMSFT reviewed (approved):
LGTM

Wwwsylvia commented on registry/remote/repository.go:
```diff
@@ -944,61 +952,132 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	if len(refDigest) > 0 && refDigest != contentDigest {
+		return ocispec.Descriptor{}, fmt.Errorf("%s: mismatch digest: %s", refDigest, contentDigest)
```
nit: You may want to make it like line 1074?

nima replied:
Good call; Done!

Wwwsylvia reviewed (approved):
LGTM with s minor suggestion

nima commented:
And a final double-check...

```
% oras pull 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Downloaded empty artifact
Pulled 080565210187.dkr.ecr.us-west-2.amazonaws.com/my.repo:my.tag
Digest: sha256:410bc64c7dbbc3286211561a0f6ca43affdca5c92c568f838a37e34640f5f76e
```

Wwwsylvia reviewed (approved):
LGTM!

gokarnm commented:
Thanks @nima for chasing this issue down, and @shizhMSFT @Wwwsylvia  for the reviews!
@shizhMSFT @FeynmanZhou do we plan to release this fix with a patch release?

shizhMSFT commented:
We plan to release this in the next rc release.

michaelb990 commented:
Nope! That’s my Java showing 😇… carry on.

On Thu, Aug 18, 2022 at 17:25 Nima Talebi ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In registry/remote/repository.go
> <https://github.com/oras-project/oras-go/pull/237#discussion_r949699977>:
>
> >  		return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
>  	}
>  	return nil
>  }
>
> +// verifyContentDigests is a superset of verifyContentDigest; it takes in the actual (calculated) digest, and
> +// verifies against that, the other two digests, both of which may or may not be set to the empty digest.  If empty,
> +// there is no validation error, however is not empty, it ensures that the given digest is valid (matches the actual
> +// digest)
> +func verifyContentDigests(actualDigest digest.Digest, digests map[string]digest.Digest) error {
> +	for src, d := range digests {
> +		if len(d.String()) == 0 {
>
> Because you can't set a string to a nil; it's not a *digest.Digest
> (pointer). Am I missing something?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oras-project/oras-go/pull/237#discussion_r949699977>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDST5SMDJGM3IW22JF4CDVZ3IBBANCNFSM53IDEHXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
